### PR TITLE
Factor out data transformations

### PR DESCRIPTION
**Pull request by jschneider-bensch**
As a first step towards a cleaner spec, that can be more closely matched to e.g. the ProVerif model, I've factored out the transformations that need to happen to pairs of identifiers/pseudonyms (I've introduced the term `handle` here) and data values, plain or encrypted.

At the moment it's a bit awkward because the table data structures don't know about these sorts of pairs yet, but that should be fixed together with #18 .

### CODE REVIEW

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Thanks, looks good with some nits.

File: hacspec-scrambledb/scrambledb/src/data_types.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    pub(crate) attribute_name: String,
+    pub(crate) value: Vec<u8>,
+}
+
+pub struct EncryptedDataValue {
+    pub(crate) attribute_name: String,
+    pub(crate) value: Vec<u8>,
+}
+
+pub struct IdentifiableDatum {
```
[the remark this answers is not in the document]
Comment (jschneider-bensch) [reply hacspec-scrambledb/scrambledb/src/data_types.rs]: My intent was to emphasize that this is about individual pieces of data, but yeah it does sound odd. I've changed it to -`Data`.

File: hacspec-scrambledb/scrambledb/src/data_transformations.rs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    let blinded_handle = BlindedIdentifiableHandle(oprf::coprf::coprf_online::blind(
+        *bpk,
+        datum.handle.as_bytes(),
+        pseudonymization_context_string(),
+        randomness,
+    )?);
+
+    // Encrypt data towards receiver
+    let HPKEConfig(_, kem, _, _) = crate::HPKE_CONF;
+    let encrypted_data_value = EncryptedDataValue {
```
[the remark this answers is not in the document]
Comment (jschneider-bensch) [reply hacspec-scrambledb/scrambledb/src/data_transformations.rs]: For now, I chose not to introduce a constructor method.

File: hacspec-scrambledb/scrambledb/src/data_transformations.rs [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+            None,
+            None,
+            None,
+            randomness.bytes(Nsk(kem))?.to_vec(),
+        )?)
+        .to_bytes(),
+    };
+
+    Ok(BlindedIdentifiableDatum {
+        handle: blinded_handle,
```
[the remark this answers is not in the document]
Comment (jschneider-bensch) [reply hacspec-scrambledb/scrambledb/src/data_transformations.rs]: I've changed the field names accordingly.

File: hacspec-scrambledb/scrambledb/src/data_transformations.rs [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+        datum.handle.as_bytes(),
+        pseudonymization_context_string(),
+        randomness,
+    )?);
+
+    // Encrypt data towards receiver
+    let HPKEConfig(_, kem, _, _) = crate::HPKE_CONF;
+    let encrypted_data_value = EncryptedDataValue {
+        attribute_name: datum.data_value.attribute_name.clone(),
+        value: SerializedHPKE::from_hpke_ct(&HpkeSeal(
```
[the remark this answers is not in the document]
Comment (jschneider-bensch) [reply hacspec-scrambledb/scrambledb/src/data_transformations.rs]: I've pulled out the double encryption and decryption into their own module.

File: hacspec-scrambledb/scrambledb/src/data_transformations.rs [hunk 5]
```diff
@@ -0,0 +1,281 @@
+use hacspec_lib::Randomness;
+use libcrux::hpke::{kem::Nsk, HPKEConfig, HpkeOpen, HpkeSeal};
+use oprf::coprf::coprf_setup::BlindingPublicKey;
+
+use crate::{data_types::*, error::Error, setup::StoreContext, SerializedHPKE};
+
+fn pseudonymization_context_string() -> Vec<u8> {
+    b"CoPRF-Context-Pseudonymization".to_vec()
```
[the remark this answers is not in the document]
Comment (jschneider-bensch) [reply hacspec-scrambledb/scrambledb/src/data_transformations.rs]: Indeed, they should be `const`s. I've changed it accordingly.

File: hacspec-scrambledb/scrambledb/src/data_transformations.rs [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+/// - bpk: Receiver's blinding public key
+/// - ek: Receiver's public encryption key
+/// - datum: Identifiable datum
+/// - randomness: Random bytes
+///
+/// Output:
+/// [BlindedIdentifiableDatum] such that the datum's handle is blinded for
+/// CoPRF evaluation and the datum's value is level-1 encrypted.
+pub fn blind_identifiable_datum(
+    bpk: &BlindingPublicKey,
```
[the remark this answers is not in the document]
Comment (jschneider-bensch) [reply hacspec-scrambledb/scrambledb/src/data_transformations.rs:39]: I'll have to make this consistent one way or the other across the spec in a follow up.

**Review by franziskuskiefer — APPROVED** [review]
Thanks!

**Pull request merged**